src/librustc_mir/util/liveness.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Liveness analysis.
+
+// FIXME: Make sure this analysis uses proper MIR semantics. Also find out what
+//        the MIR semantics are.
+
+use rustc::mir::*;
+use rustc::mir::visit::{LvalueContext, Visitor};
+use rustc_data_structures::indexed_vec::{IndexVec, Idx};
+use rustc_data_structures::indexed_set::IdxSetBuf;
+use util::pretty::{write_basic_block, dump_enabled, write_mir_intro};
+use rustc::mir::transform::MirSource;
+use rustc::ty::item_path;
+use std::path::{PathBuf, Path};
+use std::fs;
+use rustc::ty::TyCtxt;
+use std::io::{self, Write};
+
+pub type LocalSet = IdxSetBuf<Local>;
+
+#[derive(Eq, PartialEq, Clone)]
+struct BlockInfo {
+    defs: LocalSet,
+    uses: LocalSet,
+}
+
+struct BlockInfoVisitor {
+    pre_defs: LocalSet,
+    defs: LocalSet,
+    uses: LocalSet,
+}
+
+impl<'tcx> Visitor<'tcx> for BlockInfoVisitor {
+    fn visit_lvalue(&mut self,
+                    lvalue: &Lvalue<'tcx>,
+                    context: LvalueContext<'tcx>,
+                    location: Location) {
+        if let Lvalue::Local(local) = *lvalue {
+            match context {
+                LvalueContext::Store | LvalueContext::Call => {
+                    self.defs.add(&local);
+                }
+                LvalueContext::Projection(..) |
+                LvalueContext::Borrow { .. } |
+                LvalueContext::Inspect |
+                LvalueContext::Consume |
+                LvalueContext::Drop => {
+                    // Ignore uses which are already defined in this block
+                    if !self.pre_defs.contains(&local) {
+                        self.uses.add(&local);
+                    }
+                }
+                LvalueContext::StorageLive | LvalueContext::StorageDead => (),
```

Comment:
These both are defs (i.e. they kill the current value). This could make some values that are only used by drops dead. e.g.

```Rust
loop {
    let x = { // storagelive
        yield;
        Box::new(0)
    };
    // drop; storagedead
}
```